### DISCUSSION TITLE
Adding OSM data masks to an existing torchGeo geospatial data set

### DISCUSSION THREAD

**Discussion by dsterlingdg** [torchgeo/torchgeo#951] [Q&A]
I would like to create a custom data loader that wraps an existing torchGeo geospatial data set to return a data mask containing Open Street Maps (OSM) labels in addition to the default image and label mask. I'm relatively new to working with geospatial data --- are there primitives in torchGeo that I can/should use to carry out the task of acquiring the OSM data for a given bounding box and then performing any coordinate system transformation/registration tasks required to get pixel level alignment of OSM labels and the raster data?

**Comment by calebrob6** [comment]
Hi @dsterlingdg, this doesn't exist in TorchGeo (although would be a cool feature!). To implement this you would need to know how to talk to an OSM API (e.g. Overpass). The `osmnx` library does this and I have an example notebook that uses `osmnx` to rasterize roads networks from that to pixel-alignment with other imagery -- https://gist.github.com/calebrob6/5039fd409921606aa5843f8fec038c03.

**Reply by dsterlingdg** [reply]
I expected to need to learn my way around a suitable OSM API,  thanks for sharing your example @calebrob6! Let me play with this for a bit and I'll reach out if I have additional questions

**Comment by adamjstewart** [comment]
To add to what @calebrob6 said, you would want to create a new `OSM` dataset that subclasses `GeoDataset` and implement the `__getitem__` method that creates a raster given a geospatial bounding box. All other stuff (wrapping existing datasets, adding to images and label masks, CRS transforms, pixel alignment, etc.) will be taken care of automatically by TorchGeo. Once `OSM` is written, the rest of the code would look like:
```python
image = ...  # some image dataset, might be different res or CRS
mask = ...  # some mask dataset, might be different res or CRS
osm = OSM(...)

dataset = image & mask & osm
dataloader = DataLoader(dataset, ...)
```

**Reply by dsterlingdg** [reply]
Thanks @adamjstewart -- this was exactly what I had in mind. Let me see if I can cobble together the innards of an OSM dataset object.
